feat(dashboard,api-service): add self-hosted-auth

djabarovgeorge opened the pull request:
### What changed? Why was the change needed?

<!-- Also include any relevant links, such as Linear tickets, Slack discussions, or design documents. -->
- Create System User 
- Create a mock clerk library
- Auto generate self-hosted jwt token

[docs update ](https://github.com/novuhq/docs/pull/832)

### Screenshots

<!-- If the changes are visual, include screenshots or screencasts. -->

<details>
<summary><strong>Expand for optional sections</strong></summary>

### Related enterprise PR

<!-- A link to a dependent pull request  -->

### Special notes for your reviewer

<!-- Specific instructions or considerations you want to highlight for the reviewer. -->

</details>

djabarovgeorge committed:
feat(dashboard,api-service): add self-hosted-auth

scopsy committed:
fix:

scopsy committed:
fix: opted in status

djabarovgeorge committed:
Merge branch 'next' into self-hosted-auth

djabarovgeorge committed:
refactor(api-service,dashaboard): integrate SystemOrganizationService and streamline logMeIn method in AuthController, fix clerk mock

djabarovgeorge committed:
refactor(api-service): rename variable 'admins' to 'users' for clarity in SystemOrganizationService

djabarovgeorge committed:
refactor(dashboard): update self-hosted configuration and remove unused clerk mock file

djabarovgeorge committed:
refactor(dashboard): export OrganizationSwitcher from self-hosted utils and remove redundant definition

djabarovgeorge committed:
refactor(dashboard): export UserButton from self-hosted utils and clean up unused UserButton definition

djabarovgeorge committed:
refactor(dashboard, api): enhance error handling in content-templates module and update query keys with myOrganization

djabarovgeorge committed:
refactor(dashboard): implement navigation for self-hosted environments in organization list and sign-up components

djabarovgeorge committed:
refactor(dashboard): update workflow row and general settings for self-hosted environment handling

djabarovgeorge committed:
refactor(api, dashboard): enhance organization creation with API service level and improve self-hosted JWT handling

djabarovgeorge committed:
refactor(dashboard): update activity filters to use API service level instead of subscription

djabarovgeorge commented on apps/api/src/app/organization/usecases/create-organization/create-organization.usecase.ts:
```diff
@@ -11,9 +11,7 @@ import { AddMemberCommand } from '../membership/add-member/add-member.command';
 import { CreateOrganizationCommand } from './create-organization.command';
 
-@Injectable({
-  scope: Scope.REQUEST,
```
is there any specific reason we're using request scope in the app? (there other providers in this PR that i update)
it’s currently limiting me from injecting certain providers into SystemOrganizationService, since it’s a singleton, and you can’t inject a request-scoped provider into a singleton.

cc @scopsy

scopsy replied:
Yes, we need to be careful about this, when Scope is defined, we can be relying on this.someState to be only available on the request context. Vs shared with other requests. So you need to investigate the whole usecase tree to ensure we don't rely on this.somethingToKeepState

djabarovgeorge replied:
good catch, i will check and refactor the usecase if needed.

djabarovgeorge committed:
Merge branch 'next' into self-hosted-auth

djabarovgeorge committed:
fix(dashboard): validate Clerk Publishable Key only when not self-hosted

djabarovgeorge committed:
refactor(dashboard): streamline organization switcher component and enhance avatar rendering

djabarovgeorge committed:
fix(dashboard): improve tooltip content and enhance avatar animation in organization switcher

djabarovgeorge committed:
feat(dashboard): add UserAvatar component and integrate it into UserButton for enhanced user representation

djabarovgeorge committed:
feat(dashboard): enhance workflow row with tooltip for unsupported workflows

djabarovgeorge committed:
fix(dashboard): update NovuBrandingSwitch to handle self-hosted scenarios and improve messaging

djabarovgeorge committed:
fix(dashboard): update FreeTierState and NovuBrandingSwitch for self-hosted support and improve button messaging

djabarovgeorge committed:
refactor(dashboard): standardize SVG attributes and enhance SideNavigation structure for better readability

djabarovgeorge committed:
fix(dashboard): update FreeTierState component to enhance messaging and layout for better user engagement

djabarovgeorge commented on apps/dashboard/src/components/environments/environments-free-state.tsx:
```diff
@@ -80,7 +81,10 @@ export function FreeTierState() {
-        <div>
+        <div className="flex flex-col items-center gap-1">
+          <p className="text-text-soft text-label-xs mb-3 text-center">
+            Unlock more environments and enhance your productivity
```
Added new copy for free and OSS

scopsy replied:
```suggestion
            To create additional custom environments, upgrade your plan.
```

djabarovgeorge committed:
fix(dashboard): handle undefined subscription in UsageCard to prevent rendering issues

djabarovgeorge committed:
refactor(dashboard): remove unused imports in restricted-switch and organization-switcher components for cleaner code

djabarovgeorge committed:
refactor(root): remove unused imports and dependencies for cleaner code structure

merrcury committed:
feat(docker): add Dockerfile and update package.json for Docker build support

merrcury committed:
Merge branch 'self-hosted-auth' of github.com:novuhq/novu into self-hosted-auth

djabarovgeorge committed:
Merge branch 'self-hosted-auth' of github.com:novuhq/novu into self-hosted-auth

scopsy commented on apps/api/src/app/auth/auth.controller.ts:
```diff
@@ -190,4 +190,9 @@ export class AuthController {
+  @Get('/self-hosted')
+  async logMeIn() {
+    return this.systemOrganizationService.getSystemOrganizationToken();
+  }
```
We need to exclude this from NOVU_ENTERPRISE env variable

djabarovgeorge replied:
this could be fine (only available on the community edition) without any additional actions, because AuthController is used only in the CommunityAuthModule, enterprise Module uses other Controller.

scopsy commented on apps/api/src/app/organization/usecases/membership/add-member/add-member.usecase.ts:
```diff
@@ -3,9 +3,7 @@ import { MemberRepository } from '@novu/dal';
-})
+@Injectable()
 export class AddMember {
   private organizationId: string;
```
For example here it's pronblematic as we use organizationId as shared state

djabarovgeorge committed:
refactor(code-snippets): enhance server URL handling for self-hosted and EU environments

djabarovgeorge committed:
refactor(auth): rename SystemOrganizationService to CommunityEditionService and update related methods for community edition support

djabarovgeorge committed:
refactor(auth): typo

djabarovgeorge committed:
fix(auth): update COMMUNITY_EDITION_USER_EMAIL to use example domain and enhance InboxButton logic for self-hosted environments

djabarovgeorge committed:
refactor(dashboard): replace RestrictedSwitch with Popover and LinkButton for branding upgrade prompts, update self-hosted redirect logic

djabarovgeorge committed:
refactor(dashboard): update NovuBrandingSwitch to include self-hosted plan logic and enhance premium feature messaging

djabarovgeorge committed:
refactor(dashboard): enhance organization and user buttons with self-hosted upgrade links and updated messaging

djabarovgeorge committed:
Merge branch 'next' into self-hosted-auth

djabarovgeorge committed:
fix(dashboard): after next merge

djabarovgeorge committed:
Merge branch 'next' into self-hosted-auth

djabarovgeorge committed:
refactor(organization-switcher): update button styles and text for improved accessibility and consistency

djabarovgeorge committed:
refactor(dashboard): implement self-hosted redirection for welcome and settings pages

djabarovgeorge committed:
refactor(header-navigation): update customer support button to handle self-hosted redirection

djabarovgeorge committed:
refactor(user-button): enhance dropdown menu with new self-hosted upgrade options and updated icons

djabarovgeorge committed:
refactor(workflow-row): update self-hosted upgrade link to use dynamic URL for migration guide

scopsy commented on apps/api/src/app/auth/auth.controller.ts:
```diff
@@ -190,4 +190,9 @@ export class AuthController {
     return await this.authService.getSignedToken(user, organizationId, member as MemberEntity);
   }
+
+  @Get('/self-hosted')
+  async logMeIn() {
+    return this.systemOrganizationService.getCommunityEditionOrganizationToken();
+  }
```
@djabarovgeorge let's add an ENV variable here to disable when running in NOVU_ENTERPRISE

djabarovgeorge replied:
this could work out of the box because this Controller is available only in the community edition.

scopsy commented on apps/dashboard/dockerfile:
```diff
@@ -0,0 +1,26 @@
+COPY [".prettierrc",".prettierignore", "./"]
+RUN pnpm build:dashboard
+
+FROM alpine:latest
+RUN apk add --no-cache busybox
+WORKDIR /www
+COPY --from=builder /usr/src/app/apps/dashboard/build /www
+EXPOSE 80
+CMD ["busybox", "httpd", "-f", "-p", "80"]
```
@merrcury @djabarovgeorge remind, me how is the user going to modify the dashboard .env variables when starting this docker? For example changing the API_ROOT to his own domain after deploying, etc..

scopsy commented on apps/dashboard/src/components/environments/environments-free-state.tsx:
```diff
@@ -90,11 +95,16 @@ export function FreeTierState() {
             leadingIcon={RiSparkling2Line}
           >
-            Upgrade to Team Tier
+            {IS_SELF_HOSTED ? 'Migrate to Cloud' : 'Upgrade to Team Tier'}
```
```suggestion
            {IS_SELF_HOSTED ? 'Contact Sales' : 'Upgrade to Team Tier'}
```

scopsy commented on apps/dashboard/src/components/workflow-row.tsx:
```diff
@@ -198,7 +202,34 @@ export const WorkflowRow = ({ workflow }: WorkflowRowProps) => {
+            <TooltipPortal>
+              <TooltipContent side="bottom" align="center" className="z-50">
+                <div className="flex flex-col gap-1">
+                  <span className="font-medium">This workflow is not supported in this version</span>
```
```suggestion
                  <span className="font-medium">This workflow is not supported in this version of the dashboard.</span>
```

scopsy commented on apps/dashboard/src/main.tsx:
```diff
@@ -106,7 +108,7 @@ const router = createBrowserRouter([
               {
                 path: ROUTES.WELCOME,
-                element: <WelcomePage />,
+                element: IS_SELF_HOSTED ? <Navigate to={ROUTES.ROOT} replace /> : <WelcomePage />,
```
Is there a reason of why not keeping it? I think it might be helpful for onboarding, even on OSS. Wdyt?

djabarovgeorge replied:
I think we need to revisit and restructure this page, some elements only partially make sense in the current context. For example, having both “Create Account” and “Invite Member” feels a bit out of place.
I’m not against having them, but we haven’t had a proper discussion around this yet.
A cleaner approach could be to focus on more actionable, relevant items like: "Create Workflow" "Add Inbox to Your App"
![image](https://github.com/user-attachments/assets/131d9a43-c755-4b85-b3e9-916889c119a4)

Let me know what you think. Open to feedback or other suggestions on how to streamline this page.

djabarovgeorge replied:
i brought it back without the non-relevant steps, and the get started on the footer is back as well.
![image](https://github.com/user-attachments/assets/7014b475-7e99-4320-834a-b3b17e5793b6)

scopsy commented on apps/dashboard/src/utils/code-snippets.ts:
```diff
@@ -19,12 +19,18 @@ const safeParsePayload = (payload: string) => {
 
 const novu = new Novu({ 
-  secretKey: ${renderedSecretKey}${euServerUrl}
+  secretKey: ${renderedSecretKey}${serverUrl}
```
What about your fancy solution with JSON.stringift? :P

djabarovgeorge replied:
In the end, it felt more complicated! 
of of the reasons is the the `variables` would render as string, meaning:
```
secretKey : "process.env.SECRET_KEY_ENV_KEY"
```

scopsy commented on apps/dashboard/src/utils/self-hosted/components.tsx:
```diff
@@ -0,0 +1,65 @@
+    }
+  }, [navigate]);
+
+  // if (!(window as any).Clerk.loggedIn) {
+  //   console.log('RedirectToSignIn: return null');
+  //   return null;
+  // }
```
```suggestion

```

scopsy commented on apps/dashboard/src/utils/self-hosted/user.resource.tsx:
```diff
@@ -0,0 +1,31 @@
+  const jwt = localStorage.getItem('self-hosted-jwt');
+  const decodedJwt: DecodedJwt | null = jwt ? JSON.parse(atob(jwt.split('.')[1])) : null;
+  const value = {
+    user: {
+      update: async () => null,
+      externalId: decodedJwt?._id,
+      firstName: decodedJwt?.firstName,
+      lastName: decodedJwt?.lastName,
+      emailAddresses: [{ emailAddress: decodedJwt?.email }],
+      createdAt: new Date(),
+      publicMetadata: { newDashboardOptInStatus: 'opted_in' },
+      unsafeMetadata: { newDashboardOptInStatus: 'opted_in' },
+      organizationMemberships: [{}],
+      passwordEnabled: true,
+    },
```
Saw a duplication of this object somewhere else, can we combine them into a single mapped function?

djabarovgeorge committed:
refactor(auth): replace CommunityEditionService with SelfHostUsecase and remove unused method

djabarovgeorge committed:
refactor(auth): update SelfHostUsecase to use CommunityOrganizationRepository and improve dependency injection

djabarovgeorge committed:
refactor(api): simplify isMember check by passing organizationId directly

djabarovgeorge committed:
refactor(root): after pr comments

djabarovgeorge committed:
refactor(user-button): remove user email display from dropdown menu for cleaner UI

djabarovgeorge committed:
feat(dashboard): update after pr comments

djabarovgeorge committed:
Merge branch 'next' into self-hosted-auth

merrcury committed:
feat(root, dashboard): new dockerfile for the dashboard service (#8221)

djabarovgeorge committed:
fix(dashboard): remove email as user name

djabarovgeorge committed:
Merge branch 'self-hosted-auth' of github.com:novuhq/novu into self-hosted-auth

djabarovgeorge committed:
feat(api): implement self-host secret guard and update auth controller for self-hosted login

djabarovgeorge committed:
Merge branch 'next' into self-hosted-auth

djabarovgeorge committed:
refactor(dashboard): modularize demo workflow initialization and clean up unused code

- Introduced `useInitDemoWorkflow` hook to encapsulate demo workflow initialization logic.
- Removed unused imports and code related to workflow fetching in `InboxPlayground`.
- Updated `InboxConnectedGuide` to utilize the new hook for demo workflow setup.
- Cleaned up `CommunityEditionService` by removing unused member repository and switch organization use case.

djabarovgeorge committed:
Merge branch 'next' into self-hosted-auth

djabarovgeorge committed:
fix(api): adjust max snooze duration logic based on NOVU_ENTERPRISE environment variable

djabarovgeorge committed:
refactor(auth): rename self-host secret key to self-host token and update related logic

- Changed environment variable from `SELF_HOSTED_SECRET_KEY` to `SELF_HOSTED_TOKEN`.
- Updated header keys in `SelfHostSecretGuard` and `jwt-manager` to reflect the new naming.
- Adjusted error messages for missing and invalid tokens.

djabarovgeorge committed:
chore(docker): add SELF_HOSTED_TOKEN and VITE_SELF_HOSTED_TOKEN to docker-compose.yml

djabarovgeorge committed:
fix(auth): update error messages in SelfHostSecretGuard for clarity

- Changed error message for missing self-host token to 'Missing self-hosted token'.
- Enhanced invalid token error message to include guidance on validating the API and Dashboard token consistency.

djabarovgeorge committed:
fix(dashboard): add bell for test page

- Simplified visibility conditions for the InboxButton component.
- Removed unnecessary check for IS_SELF_HOSTED when user and currentEnvironment are not present.
- Ensured the button only renders in appropriate contexts, improving clarity and performance.

djabarovgeorge committed:
Merge branch 'next' into self-hosted-auth

djabarovgeorge committed:
Merge branch 'next' into self-hosted-auth

djabarovgeorge committed:
fix(dashboard): remove TypeScript error suppression in maily-config

- Eliminated unnecessary @ts-expect-error comments in the createExtensions function for better type safety and clarity.

djabarovgeorge committed:
fix(api-service): ci run

- Removed CommunityEditionService from providers in community.auth.module.config.ts for cleaner module configuration.
- Enhanced link-entities.service.spec.ts by adding LoggerModule for improved logging capabilities during tests.
- Updated session.usecase.ts to include additional condition for maxSnoozeDurationHours based on CI_EE_TEST environment variable.

djabarovgeorge committed:
fix(api): update test command and simplify maxSnoozeDurationHours logic

- Updated the test command in package.json to include NOVU_ENTERPRISE and CLERK_ENABLED environment variables for improved testing context.
- Simplified the logic for maxSnoozeDurationHours in session.usecase.ts by removing the CI_EE_TEST condition, enhancing code clarity.

djabarovgeorge committed:
Merge branch 'next' into self-hosted-auth

djabarovgeorge committed:
refactor(api): replace hardcoded COMMUNITY_EDITION_NAME with helper function

- Updated CommunityEditionService and SelfHostUsecase to use getSelfHostedFindQuery for organization retrieval instead of hardcoded COMMUNITY_EDITION_NAME.
- This change improves maintainability and reduces duplication of the organization name logic.

djabarovgeorge committed:
fix(api-service): ci test

djabarovgeorge committed:
fix(api-service): ci tests
